### ISSUE TITLE
`split_dataset` cannot split a dataset containing nested tuples

### ISSUE THREAD

**Issue by Aweptimum** [keras-team/keras#19797]
I'm using tensorflow and loading data into a `tf.data.Dataset` object structured like so, where each dataset has the same length (20160 entries in this case):
```python
data1 = tf.data.Dataset(...) # tensorspec of [1, 16384, 2]
data2 = tf.data.Dataset(...) # tensorspec of [512, 32, 2]
labels = tf.data.Dataset(...) # tensorspec of scalar int32

nested = (data1, data2)
dataset = tf.data.Dataset.zip(nested, labels)
```
This is to comply with the input to [`model.fit`](https://www.tensorflow.org/api_docs/python/tf/keras/Model#fit):
> A [tf.data.Dataset](https://www.tensorflow.org/api_docs/python/tf/data/Dataset). Should return a tuple of either (inputs, targets) or (inputs, targets, sample_weights)

However, when trying to split into training and validation sets, `_restore_dataset_from_list` throws an error. Specifically, this line:
https://github.com/keras-team/keras/blob/26f438edd5516bf02504a5a0e13a20ceec69cb29/keras/src/utils/dataset_utils.py#L455

It's hard for me to see what's happening exactly because `np.array` is what's throwing the error:
```
setting an array element with a sequence. The requested array has an inhomogeneous shape after 2 dimensions.
The detected shape was (20160, 2) + inhomogeneous part
```

But this seems to indicate the inhomogenous part is the label, and if it's trying to coerce the structure I think it is (below) then it would  indeed be a violation of the allowed structure for a numpy array
```python
[
    [[d1, d2], label]
    ...
]
```
Given that a `tf.data.Dataset` explicitly supports nesting and `model.fit` expects a tuple of inputs for a multi-input model, I think `split_dataset` should be able to handle nesting as well. But if handling nesting requires a slower method, then the current method should be kept and nesting handled as a fallback in case `np.array` fails.

My workaround for the time being is to zip a flat dataset and then pack the inputs *after* the call to `split_dataset` using `map`:
```python
dataset = tf.data.Dataset.zip(data1, data2, labels)
trainer, valer = tf.keras.utils.split_dataset(dataset, 0.7)
trainer = trainer.map(lambda a,b,l: ((a,b), l))
valer   =   valer.map(lambda a,b,l: ((a,b), l))
```
Which isn't awful, but it took me a very long time to figure out the issue. I'd just like to save others the headache :)

Edit: or does the label dataset need to be a list of 1D tensors with single elements instead of scalars?

**Comment by Aweptimum** [comment]
@hertschuh thanks a bunch! You were fast 😄
